rust/krb5-provision-keytab/src/main.rs

```diff
@@ -109,20 +157,136 @@ fn run() -> Result<Response, Error> {
+                    }
+                    res => res.context(CreatePrincipalSnafu { principal: &princ })?,
+                }
+                let keys = kadmin
+                    .get_principal_keys(&princ, KVNO_ALL)
+                    .context(GetPrincipalKeysSnafu { principal: &princ })?;
+                for key in keys.keys() {
+                    kt.add(&princ, key.kvno, &key.keyblock)
+                        .context(AddToKeytabSnafu { principal: &princ })?;
+                }
+            }
+            AdminConnection::ActiveDirectory {
+                ldap,
+                password_cache_secret,
+                user_distinguished_name,
+                schema_distinguished_name,
+            } => {
+                let kube = stackable_operator::client::create_client(None)
+                    .await
+                    .unwrap();
+                let princ_name = princ.to_string();
+                let mut password_cache = kube
+                    .get::<Secret>(
+                        password_cache_secret.name.as_deref().unwrap(),
+                        password_cache_secret.namespace.as_deref().unwrap(),
+                    )
+                    .await
+                    .unwrap();
+                let password_cache_key = princ_name.replace(['/', '@'], "--");
+                let password = if let Some(pw) = password_cache
+                    .data
+                    .get_or_insert_with(BTreeMap::default)
+                    .get(&password_cache_key)
+                {
+                    tracing::info!(
+                        principal = princ_req.name,
+                        cache_key = password_cache_key,
+                        "found principal in key cache, reusing"
+                    );
+                    pw.0.clone()
+                } else {
+                    tracing::info!(
+                        principal = princ_req.name,
+                        cache_key = password_cache_key,
+                        "did not find principal in key cache, creating"
+                    );
+                    tracing::info!(principal = ?princ_req.name, "Creating principal");
+                    let principal_cn = ldap3::dn_escape(&*princ_req.name);
+                    // FIXME: AD restricts RDNs to 64 characters
+                    let principal_cn = principal_cn.get(..64).unwrap_or(&*principal_cn);
+                    let password = generate_ad_password(40);
+                    let password_ad_encoded = {
+                        let mut pwd_utf16le = Vec::new();
+                        format!("\"{password}\"").encode_utf16().for_each(|word| {
+                            WriteBytesExt::write_u16::<LittleEndian>(&mut pwd_utf16le, word)
+                                .unwrap()
+                        });
+                        pwd_utf16le
+                    };
+                    let user_dn = format!("CN={principal_cn},{user_distinguished_name}",);
+                    ldap.add(
```

Comment:
Can we please have at least a descriptive error message including which user it is trying to create, so that e.g. already existent users can be identified

Reply:
thx!